Joseph Caguioa

Spring 2020

DS 7337: Natural Language Processing

Section 404 (Tuesday 2030-2200)

HW5 Due: Date of Live Session 9 (3/3/20)

---

# Homework 5

## <u><a name="toc">Table of Contents:</a></u>
* [Question 1](#question1)
* [Question 2](#question2)
* [Question 3](#question3)

---

### <a name="question1">Question 1</a> 

<b>Compile a list of static links (permalinks) to individual user movie reviews from one particular website. This will be your working dataset for this assignment, as well as for assignments 7 and 8, which together will make up your semester project.  
  
* It does not matter if you use a crawler or if you manually collect the links, but you will need at least 100 movie review links. Note that, as of this writing, the robots.txt file of IMDB.com allows the crawling of user reviews.
* Each link should be to a web page that has only one user review of only one movie, e.g., the user review permalinks on the IMDB site.
* Choose reviews of movies that are all in the same genre, e.g., sci-fi, mystery, romance, superhero, etc.
* Make sure your collection includes reviews of several movies in your chosen genre and that it includes a mix of negative and positive reviews.</b> <sub>[(back to top)](#toc)</sub>

In [1]:
review_permalinks = [
    # Rocky (1976)
    'https://www.imdb.com/review/rw0149681/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0149646/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0995951/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0149549/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0149655/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3439530/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0149580/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3548992/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2455895/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4349307/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2740261/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0149534/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1701319/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0149538/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1979442/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2406266/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3006171/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3315551/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3218345/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3794060/?ref_=tt_urv',
    # Rocky II (1979)
    'https://www.imdb.com/review/rw0170478/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0170493/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2008119/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1192366/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3649962/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0170457/?ref_=tt_urv',
    'https://www.imdb.com/review/rw5154437/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2405186/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1216821/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1179397/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2250841/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2180823/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1779351/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3385719/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3218720/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2216519/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1320829/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3652552/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3829829/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3384789/?ref_=tt_urv',
    # Rocky III (1982)
    'https://www.imdb.com/review/rw3348405/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1563301/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2033895/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1068592/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1761792/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1378407/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0196275/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3785443/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3608003/?ref_=tt_urv',
    'https://www.imdb.com/review/rw5156778/?ref_=tt_urv',
    'https://www.imdb.com/review/rw5367975/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2037512/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1320840/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1765787/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3322668/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4556877/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3218837/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3598350/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0196331/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1779344/?ref_=tt_urv',
    # Rocky IV (1985)
    'https://www.imdb.com/review/rw3348827/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1810997/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1400935/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4228386/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1513871/?ref_=tt_urv',
    'https://www.imdb.com/review/rw5034284/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1412571/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4102161/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0226208/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1644245/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0226256/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0226271/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1904244/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0226205/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2271803/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3091967/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0226278/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2361152/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1303506/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0226242/?ref_=tt_urv',
    # Rocky V (1990)
    'https://www.imdb.com/review/rw1341699/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2035169/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2763282/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1845257/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1521425/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1553738/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2557900/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1555987/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1288277/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1539376/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4991605/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2083896/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3069103/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1047005/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2633512/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0287686/?ref_=tt_urv',
    'https://www.imdb.com/review/rw0287677/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1320858/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1388725/?ref_=tt_urv',
    'https://www.imdb.com/review/rw2023052/?ref_=tt_urv',
    # Rocky Balboa (2006)
    'https://www.imdb.com/review/rw1547658/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1541991/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1547649/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1542029/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1506437/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1536601/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3348786/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1536660/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1547666/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1541992/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1825591/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1560341/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1623005/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1608355/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1604980/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1563462/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1627901/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1625520/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1618049/?ref_=tt_urv',
    'https://www.imdb.com/review/rw1555952/?ref_=tt_urv',
    # Creed (2015)
    'https://www.imdb.com/review/rw3360962/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3361216/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3361234/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3438067/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3361248/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3443782/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3361280/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3390705/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3382852/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3406176/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4111703/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3386188/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3380641/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3605637/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3376293/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3365803/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3381785/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3396436/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3369297/?ref_=tt_urv',
    'https://www.imdb.com/review/rw3379530/?ref_=tt_urv',
    # Creed II (2018)
    'https://www.imdb.com/review/rw4513714/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4470739/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4467731/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4477268/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4484837/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4476894/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4470467/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4469830/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4484631/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4470656/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4502055/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4493475/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4513279/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4492430/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4642148/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4643111/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4528823/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4475849/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4657962/?ref_=tt_urv',
    'https://www.imdb.com/review/rw4506429/?ref_=tt_urv'
]

---

### <a name="question2">Question 2</a> 

<b>Extract noun phrase (NP) chunks from your reviews using the following procedure:

* In Python, use BeautifulSoup to grab the main review text from each link.  
* Next run each review text through a tokenizer, and then try to NP-chunk it with a shallow parser. 
* You probably will have too many unknown words, owing to proper names of characters, actors, and so on that are not in your working dictionary. Make sure the main names that are relevant to the movies in your collection of reviews are added to the working lexicon, and then run the NP chunker again.</b> <sub>[(back to top)](#toc)</sub>

In [2]:
from bs4 import BeautifulSoup
import requests
import json
import nltk

def parse_html(url):
    page = requests.get(url)
    soup = BeautifulSoup(page.content, 'html.parser')
    return soup

# Unused due to need for try/except loop on pages without 'reviewBody'
def review_contents(soup):
    json_object = soup.select('script[type*="application/ld+json"]')
    reviewDict = json.loads(str(json_object[0].get_text()))
    movie = reviewDict['itemReviewed']['url']
    author = reviewDict['author']['name']
    rating = reviewDict['reviewRating']['ratingValue']
    review = reviewDict['reviewBody']
    return review

In [3]:
review_texts = []

for link in review_permalinks:
    soup = parse_html(link)
    json_object = soup.select('script[type*="application/ld+json"]')
    reviewDict = json.loads(str(json_object[0].get_text()))
    try:
        review = reviewDict['reviewBody']
    except KeyError:
        continue
    review_texts.append(review)

review_texts

['When I was a kid there were only three major things in my life: food, wrestling and `Rocky\'. My brother and I followed the numerous Rocky Balboa struggles religiously, and even today, there is no exception. Although the latest efforts of Stallone have been pitiful, the movie that made his career is the 1977 Academy Award Winner for Best Picture, `Rocky\'. On top of that, `Rocky\' also earned two more Academy Awards for Best Film-Editing and Best Director (John G. Avildsen).\n\nThis movie is unlike any other ever put on film. It\'s films like this that remind us of the beauty that is going to the cinema, and if only I was around during that year when such films as `Close Encounters of the Third Kind\', `Jaws\', `Saturday Night Fever\' and `Network\' were all playing in cinemas. The storyline takes place in Philadelphia, where Rocky Balboa (Sylvester Stallone) struggles to make a living as both a small-time boxer, and the brawn of a loan shark. Until one day, a sudden opportunity is h

In [4]:
len(review_texts)

116

In [5]:
reviews_tokenized = [nltk.word_tokenize(review) for review in review_texts]   
len(reviews_tokenized)

116

In [6]:
import spacy
import en_core_web_sm
eng_mod = en_core_web_sm.load()
nlp = spacy.load("en_core_web_sm")

In [7]:
review_docs = [nlp(review) for review in review_texts]

review_chunks = []

i = 0
while i < len(review_docs):
    review_chunks.append([chunk.text for chunk in review_docs[i].noun_chunks])
    i += 1

In [8]:
%pprint

review_chunks

Pretty printing has been turned OFF


[['I', 'a kid', 'only three major things', 'my life', 'food', 'wrestling', '`Rocky', 'My brother', 'I', 'Rocky Balboa', 'no exception', 'the latest efforts', 'Stallone', 'his career', 'the 1977 Academy Award Winner', 'Best Picture', 'Rocky', 'top', 'Rocky', 'two more Academy Awards', 'Best Film-Editing', 'Best Director', 'John G. Avildsen', 'This movie', 'film', 'It', 'films', 'us', 'the beauty', 'the cinema', 'I', 'that year', 'such films', '`Close Encounters', 'the Third Kind', '`Jaws', '`Saturday Night Fever', '`Network', 'cinemas', 'The storyline', 'place', 'Philadelphia', 'Rocky Balboa', 'Sylvester Stallone', 'a living', 'both a small-time boxer', 'a loan shark', 'one day', 'a sudden opportunity', 'Rocky', 'the World Heavy-Weight Title', 'the champion', 'Apollo Creed', 'Carl Weathers', 'a charismatic and flamboyant fighter', '`The King', 'Sting', '`The Master', 'Disaster', "Rocky's reputation", '"The Italian Stallion', 'a match', 'Creed', 'the unknown boxer', 'a fight', 'a "nobody

---

### <a name="question3">Question 3</a> 

<b>Output all the chunks in a single list for each review, and submit that output for this assignment. Also submit a brief written summary of what you did (describe your selection of genre, your source of reviews, how many you collected, and by what means).</b> <sub>[(back to top)](#toc)</sub>

For continuity and potential ease of recognition of entities, movies from the same series are used. Specifically, the current eight movies of the <em>Rocky</em> series available as of March 2020. All entries under the <em>Rocky</em> and <em>Creed</em> titles can be considered sports drama films.

IMDB is the source of reviews. Attempts were made to write a crawler before I gave up and manually collected permalinks to individual reviews. I collected ten positive and ten negative reviews (generally classified as at above or below a 5/10 rating, respectively) for each title; a total of 160 links are in the list in Q1. However, my BeautifulSoup code that relied on review text being present in a JSON object in the head of individual pages unexpectedly fell apart on reviews where the text is initially hidden by a spoiler warning. Rather than spend more time replacing those, I set my collector of review texts to simply ignore pages where it could not find the JSON tag I initially wrote for. Thankfully, I was still left with 116 valid reviews.

The default shallow parser shown in the textbooks, using pattern.en, seems to only work with Python 2. I had trouble getting my Anaconda environment to change to a Python 2 kernel, so I instead use a similar parser available in spaCy. All chunk lists are available as one list of lists, review_chunks, shown above at the end of Q2. Both positive- and negative-toned instances are demonstrated below.

In [9]:
len(review_chunks)

116

In [10]:
review_chunks[1]

['I', 'I', 'this film', 'a long time', 'I', 'the second', 'the series', 'I', 'I', 'the first movie', 'I', 'it', 'the series', 'the best movies', 'Rocky', 'just a boxing movie', 'It', 'a movie', 'the unbreakable human spirit', 'determination', 'the will', 'the distance', 'Rocky(Sylvester Stallone', 'a second rate club-fighter', 'a debt collector', 'a loan shark(Joe Spinnell', 'He', 'a shy girl', 'a petshop', 'friends', 'her brother', 'Paulie(Burt Young', "Rocky's life", 'anything', 'interest', 'the World Heavyweight Champion Apollo Creed(Carl Weathers', 'him', 'an unexpected shot', 'the title', 'An old boxer-turned-trainer', "Rocky's manager", 'the whole film', 'what', 'the greatest and most memorable movie moments', 'all time', 'Rocky', 'the distance', 'the training scene', 'the final fight', 'the closing moments', 'the film', 'they', 'the human spirit', 'the will', 'anybody', 'an excellent music score', 'you', 'a classic and unforgettable movie', 'a doubt', 'the Best Picture award', '

In [11]:
review_chunks[115]

['unoriginal and predictable sequels', 'Sylvester Stallone', 'the Rocky Series', 'I', 'the Rocky movies upto', 'all the ones', 'The character', 'Adonis Creed', 'the emotional intelligence', 'a 2 year old toddler', 'The film', 'a complete rehash', 'virtually every Rocky movie', 'it', 'no value', 'anyone', 'who', 'the previous films', 'It', 'I', 'this review', 'the film', 'the last 30 minutes', 'I', 'what', 'the characters', 'You', 'no emotional attachment', 'any level', 'the main protagonists', 'anyone', 'this film', 'it', 'a punch', 'a knockout', 'other boxing pub', 'me', 'I', 'find', 'I', 'our something', 'Even the boxing scenes', 'a high quality feature', 'the Rocky movies', 'this movie', 'Terrible choreography', 'scenes', 'the cherry', 'top', 'what', 'the worst boxing movies', 'the screens', 'a long time', 'Slyvester', 'these terrible movies']